### PR TITLE
create Address message with SocketAddr, get SocketAddr from Address m…

### PR DESCRIPTION
**Pull request by tamasblummer**
…essage.

Creating addr message for a socket or socket from an addr message is a common task upstream. It makes sense to support in this base library.

### CODE REVIEW

**Comment by apoelstra** [comment]
cc @TheBlueMatt can you review this?

If not I can do it this week.

File: src/network/address.rs [hunk 1]
```diff
@@ -118,5 +146,16 @@ mod test {
                              0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff, 0x0a, 0, 0, 1]);
         assert!(addr.is_err());
     }
+
+    #[test]
+    fn test_socket_addr () {
+        let s4 = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(111,222,123,4)), 5555);
+        let a4 = Address::new(&s4, 9);
```
Comment (TheBlueMatt) [inline src/network/address.rs]: Can you add a test that this mapps correctly to bitcoin's ipv6-mapped addresses? I assume its correct, but good to check as IIRC there are multiple ways to map (I belive Bitcoin uses the ::ffff:IPv4 one, but should double-check).

File: src/network/address.rs [hunk 2]
```diff
@@ -33,6 +34,32 @@ pub struct Address {
...
+    pub fn new (socket :&SocketAddr, services: u64) -> Address {
+        let (address, port) = match socket {
+            &SocketAddr::V4(ref addr) => (addr.ip().to_ipv6_mapped().segments(), addr.port()),
+            &SocketAddr::V6(ref addr) => (addr.ip().segments(), addr.port())
+        };
+        Address { address, port, services }
+    }
+
+    /// extract socket address from an address message
+    pub fn socket_addr (&self) -> SocketAddr {
```
Comment (TheBlueMatt) [inline src/network/address.rs]: You should have a check that the contained addr isnt a Tor-wrapped one and return a Result<>, sadly, I think.

File: src/network/address.rs (hunk 1 above)
Comment (tamasblummer) [reply src/network/address.rs]: That is what Ipv6Addr::to_ipv4 checks and I used in Address::socket_addr() therefore the test should never fail.

File: src/network/address.rs (hunk 2 above)
Comment (tamasblummer) [reply src/network/address.rs]: Please give me a pointer how that is recognized. I did not yet deal with tor at this level.

**Comment by tamasblummer** [comment]
addressed comments and squashed commits

File: src/network/address.rs (hunk 1 above)
Comment (tamasblummer) [reply src/network/address.rs]: see in deserialize_address_test

**Comment by tamasblummer** [comment]
@apoelstra please remove rust 1.14 check if there is no good reason for it.

**Comment by TheBlueMatt** [comment]
The Rust 1.14 check is what's required to support Debian stable, which a lot of folks use.

**Comment by tamasblummer** [comment]
updated for rust 1.14 squashed.

**Pull request merged**